pkg/controller/deployment/deployment_controller.go

```diff
@@ -564,8 +565,62 @@ func (dc *DeploymentController) syncRollingUpdateDeployment(deployment extension
 
 	// Sync deployment status
 	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
+}
+
+// proportionalScaling should run only on scaling events and not during the normal rollout process.
+func (dc *DeploymentController) proportionalScaling(deployment extensions.Deployment) error {
+	newRS, oldRSs, err := dc.getNewAndOldReplicaSets(deployment)
+	if err != nil {
+		return err
+	}
+	if len(oldRSs) == 0 {
+		// No old rcs with pods; scale new rc to the full count.
+		_, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment)
+		return err
+	}
+	allRSs := append(oldRSs, newRS)
+
+	sort.Sort(controller.ByAvailableReplicas(allRSs))
+
+	maybeMaxSurge := 0
+	if deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType {
+		// TODO: Pretty common code that soon will be extracted into an utility
+		maybeMaxSurge, isPercent, err := util.GetIntOrPercentValue(&deployment.Spec.Strategy.RollingUpdate.MaxSurge)
+		if err != nil {
+			return fmt.Errorf("invalid value for MaxSurge: %v", err)
+		}
+		if isPercent {
+			maybeMaxSurge = util.GetValueFromPercent(maybeMaxSurge, deployment.Spec.Replicas)
+		}
+	}
 
-	// TODO: raise an event, neither scaled up nor down.
+	// Number of pods that can be either added or removed from the total pod count.
+	total := deployment.Spec.Replicas + maybeMaxSurge - deployment.Status.AvailableReplicas
```

Comment:
You are right about status.replicas. Regarding syncing status, unless something external to the controller has scaled any of the replica sets, I think we should be ok (since all the scaling should be - and currently is - coordinated by the controller). That being said, we have all the replica sets at hand, so we can just inspect those instead of d.status. Updated